### PR TITLE
[ML] Refactor data frame analytics framework in steps

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit improves the design of the code that manages
and runs data frame analytics jobs.

First, it splits the code into asynchronous steps. At the moment,
there are only two steps: the reindexing step and the analysis
step. However, splitting the task into steps allows us to
factor out running inference into its own step which in turn
allows us to properly resume a job that failed during inference
without having to start a c++ process. I will follow with this
improvement in a follow up PR.

The other main improvement this commit does is that it simplifies
the state of a DFA task by getting rid of `reindexing` and `analyzing`
states. Now, once the task goes to `started` it stays there until
it finishes or gets stopped. The removed states are no longer useful.
They used to be useful in order to know how to resume a job before
progress was added. But currently they serve no purpose at all.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java [hunk 1]
```diff
@@ -174,40 +126,56 @@ private void createStatsIndexAndUpdateMappingsIfNecessary(Client client, Cluster
...
+                        executeStep(task, config, new AnalysisStep(client, task, auditor, config, processManager));
+                        break;
+                    case ANALYSIS:
+                        // This is the last step
+                        LOGGER.info("[{}] Marking task completed", config.getId());
+                        task.markAsCompleted();
+                        break;
+                    default:
+                        task.markAsFailed(ExceptionsHelper.serverError("Unknown step [{}]", step));
+                }
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java:171]: it almost seems to me that the `stepResponse` should have this info, or maybe the step itself? Doing a switch case here seems to be missing helpful polymorphism

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/dataframe/DataFrameAnalyticsManager.java:171]: My thinking was that each step is isolated and doesn't know about the other steps. Then the manager's responsibility is to execute the steps in the right order. So this is what is done here. Depending on which step just completed, the manager starts the next appropriate step.

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Review by benwtrent — APPROVED** [review]
I like the redesign. It separates out the state machine tasks from the management of them tasks. 

Keeping mutable interactions with the persistent task to a minimum helps with reasoning as well as there are not 10s of stopping paths/etc.

**Pull request merged**